### PR TITLE
[BE] update type annotations for basic utilities in `torch/__init__.py`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Changes:

1. Make some arguments positional-only as we only support Python 3.8+
2. Clean up `torch.typename(obj)` implementation.
3. Update type annotations., especially `is_tensor()` and `is_masked_tensor()` using `TypeGuard`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -22,7 +22,9 @@
 import sys
 import textwrap
 import threading
-from typing import Any, Callable, Dict, Optional, Set, Tuple, Type, TYPE_CHECKING, Union
+import typing
+from typing import Any, Callable, Dict, Optional, Set, Tuple, TYPE_CHECKING, Union
+from typing_extensions import TypeGuard
```
Comment (XuehaiPan) [inline torch/__init__.py:27]: `typing.Type` is shadowed by `torch._C.Type`.

File: torch/__init__.py [hunk 2]
```diff
@@ -22,7 +22,9 @@
 import sys
 import textwrap
 import threading
-from typing import Any, Callable, Dict, Optional, Set, Tuple, Type, TYPE_CHECKING, Union
+import typing
+from typing import Any, Callable, Dict, Optional, Set, Tuple, TYPE_CHECKING, Union
```
Comment (malfet) [inline torch/__init__.py:26]: Why not use this one?
```suggestion
from typing import Any, Callable, Dict, Optional, Set, Tuple, Type as _Type, TYPE_CHECKING, Union
```

**Review by malfet — APPROVED** [review]
LGTM, though please consider just importing `typing.Type as `_Type` (Or `tType`)

File: torch/__init__.py (hunk 2 above)
Comment (XuehaiPan) [reply torch/__init__.py]: Thanks for the suggestions. Updated.

**Comment by malfet** [comment]
Feel free to ignore back compat(as arg rename is false positive). Changes still look OK, but please update the title to reflect why did you have to change more file rather than just init.py

**Pull request closed without merging**